File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        // time complexity -> O(m+n)
        
//         int n=matrix.size();
//         int m=matrix[0].size();
        
//         int j=m-1;
//         int i=0;
        
//         while(i<n && j>=0){
            
//             if(matrix[i][j] == target){
//                 return true;
//             }
            
//             else if(matrix[i][j] > target){
//                 j--;
//             }
//             else if(matrix[i][j] < target){
//                 i++;
//             }
//         }
        
//         return false;
        
        int n=matrix.size();
        int m=matrix[0].size();
        
        int lo=0;
        int hi=n*m-1;
        
        while(lo<=hi){
            
            int mid = lo + (hi-lo)/2;
            
            int r = mid/m;
            int c = mid%m;
            
            if(matrix[r][c] == target) return true;
            else if(matrix[r][c] > target) hi--;
            else lo++;
        }
        
        return false;
    }
};

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        n = len(matrix)
        m = len(matrix[0])
        
        start, end = 0, n*m - 1
        
        while start <= end:
            
            mid = (start + end)//2
            
            row, col = mid//m, mid % m
            
            if matrix[row][col] == target:
                return True
            
            if matrix[row][col] > target:
                end = mid - 1
            else:
                start = mid + 1
        return False